### DISCUSSION TITLE
Migrate from express to fastify

### DISCUSSION THREAD

**Discussion by LasseRosenow** [directus/directus#16958] [Feature Requests]
The biggest issue I have with directus is, that it uses soo soo much memory.
The best solution to fix this would be to rewrite the API in rust, which could save around 90% compared to the current implementation.
But since this probably to much to ask for, I hope moving to a more efficient rest server implementation could be a good alternative.

As you can see here: https://medium.com/deno-the-complete-reference/express-vs-fastify-hello-world-use-case-eca4539211b0

Fastify is much superior to express in terms of memory efficiency but also overall performance.

Also most importantly while express.js is not officially deprecated, it is not very much in development anymore. More like maintenance with only a few commits every few weeks on github. Compared to the much more active fastify or koa projects. Koa being the official successor of express, but fastify beats koa also. So I would suggest just moving to fastify.

**Comment by rijkvanzanten** [comment]
> The best solution to fix this would be to rewrite the API in rust

No big deal! 😅 

> As you can see here: https://medium.com/deno-the-complete-reference/express-vs-fastify-hello-world-use-case-eca4539211b0
> 
> Fastify is much superior to express in terms of memory efficiency but also overall performance.

This doesn't show the whole picture though.. This is comparing the two web frameworks when they're not doing anything. A 25% memory decrease in express vs fastify does not mean that directus' memory will go down 25% when we switch over. To give a silly yet not unrealistic example; by being able to handle more requests simultaneously, we might end up using more memory in the given timeframe, as more things are IOed to the database 🙂 

Fastify is great, but also comes with some opinions that make it a little tricky to adapt for some of our more specific uses. It also doesn't have the vast vast vast library of community libraries / tutorials / media etc available yet, so the learning curve can be a bit harder at times.

While I do agree that the upgrade to fastify is a thing we might want to look into in the near-future, I don't think it's as black and white as "express old, new thing better" 😁

**Comment by paescuj** [comment]
> The biggest issue I have with directus is, that it uses soo soo much memory. [...]

Serious question: Did the memory usage cause any problems or is it just something that is bothering you?

**Reply by rijkvanzanten** [reply]
> > The biggest issue I have with directus is, that it uses soo soo much memory. [...]
> 
> 
> 
> Serious question: Did the memory usage cause any problems or is it just something that is bothering you?
> 
> 

There's also the typical node confusion between reported and free memory. The v8 engine will over allocate to optimize performance, but can run with way less if less is available

**Reply by LasseRosenow** [reply]
> Serious question: Did the memory usage cause any problems or is it just something that is bothering you?

For a small company like ours memory is expensive and a server can also only be upgraded so much until we need to buy an additional server rack, that is why we are always interested in having services use as less ram as possible :)

**Reply by paescuj** [reply]
Sure, that makes sense!

Obviously, there is always room for improvement but I'm just a bit surprised, because I haven't noticed anything to that effect so far.
Might not be a super legitimate comparison, but as a reference my test instance with a larger database still shows a consumption around ~160MB after executing several fast requests. 
I wonder what it looks like in your case?

**Reply by LasseRosenow** [reply]
I don't have easy access to my K8s right now, as I am at the university right now and some ports are blocked after a massive hack that destroyed all the university systems haha, but I remember I saw one service using up to 500MB a few month ago.

I would check on that later when I am back home :)

**Reply by paescuj** [reply]
> a massive hack that destroyed all the university systems haha

Oh noooo 👨🏻‍💻👾

> but I remember I saw one service using up to 500MB a few month ago.

Okay, maybe consumption has dropped a bit over the last few versions...

> I would check on that later when I am back home :)

Great, thanks!

**Reply by LasseRosenow** [reply]
> Okay, maybe consumption has dropped a bit over the last few versions...

That would be really cool, if I find one that consumes a lot of RAM, I would try updating it also :)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
